### PR TITLE
8321371: SpinPause() not implemented for bsd_aarch64/macOS

### PR DESCRIPTION
**Pull request by fbredber**
The SpinPause() function only returns 0 on bsd_aarch64 (i.e. macOS)

This PR implements SpinPause() for MacOS on AArch64 and makes it possible to choose between none, nop, isb and yield by using the OnSpinWaitInst option. The same functionality is found on AArch64 based Linux platforms.

Tested successfully on macosx-aarch64 tier1-tier5.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8321371](https://bugs.openjdk.org/browse/JDK-8321371): SpinPause() not implemented for bsd_aarch64/macOS (**Enhancement** - P4)

### Reviewers
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to [addf22d0](https://git.openjdk.org/jdk/pull/16994/files/addf22d0d3d1ad2f920737e8db0dd368a353498a)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [159244bd](https://git.openjdk.org/jdk/pull/16994/files/159244bd02873420627cdec87863231aca73c3bf)
 * [Daniel D. Daugherty](https://openjdk.org/census#dcubed) (@dcubed-ojdk - **Reviewer**)
 * [Evgeny Astigeevich](https://openjdk.org/census#eastigeevich) (@eastig - Committer)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16994/head:pull/16994` \
`$ git checkout pull/16994`

Update a local copy of the PR: \
`$ git checkout pull/16994` \
`$ git pull https://git.openjdk.org/jdk.git pull/16994/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16994`

View PR using the GUI difftool: \
`$ git pr show -t 16994`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16994.diff">https://git.openjdk.org/jdk/pull/16994.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16994#issuecomment-1850013268)

### CODE REVIEW

**Review by fisk — APPROVED** [review]
Looks good.

**Review by dholmes-ora — APPROVED** [review]
Seems reasonable to me, but we really need the Aarch64 folk to chime in on this one.

Thanks

**Comment by shipilev** [comment]
Pinging @eastig and @theRealAph here.

**Comment by eastig** [comment]
Hi @fbredber,

If you want to implement `SpinPause` for  bsd_aarch64/macOS, you need:
- Copy the implementation of `SpinPause` from https://github.com/openjdk/jdk/blob/master/src/hotspot/os_cpu/linux_aarch64/os_linux_aarch64.cpp#L412 to your `SpinPause`. We can consider to extract it and to put it in cpu/aarch64, e.g. in https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/spin_wait_aarch64.hpp.
- In [VM_Version::initialize](https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/vm_version_aarch64.cpp#L70), you need to initialize `OnSpinWaitInst` and `OnSpinWaitInstCount` for Apple M CPUs. If you wan to use one `yield`,  the code would be:
```
    if (FLAG_IS_DEFAULT(OnSpinWaitInst)) {
      FLAG_SET_DEFAULT(OnSpinWaitInst, "yield");
    }

    if (FLAG_IS_DEFAULT(OnSpinWaitInstCount)) {
      FLAG_SET_DEFAULT(OnSpinWaitInstCount, 1);
    }
```

**Comment by eastig** [comment]
Based on `https://github.com/dougallj/applecpu`, `yield` is `nop`. I'd use one `isb` as we use it for Neoverse. On M1, `isb` is ~25-30 clocks.

**Comment by eastig** [comment]
The better way to find out an instruction to use is to run microbenchmarks/benchmarks. See #6803.

**Comment by fbredber** [comment]
Hi @eastig,

I initially did the things you suggest, but after some internal discussions changed the implementation into a single yield instruction.
See JDK-8321371 for more background information.

**Comment by eastig** [comment]
@fbredber I comments on JDK-8321371.
I see misunderstanding of the purpose of `SpinPause`:

> After having some internal discussions, it seems like the most reasonable thing to do is to implement SpinPause() using a single inline yield instruction. This way we get rid of both the call and the WX stuff.
>
> This solution also showed better performance figures than the OnSpinWaitInst options did.
>
> The reason for using the yield instruction instead of the the isb instruction (which showed slightly better performance figures) is that the yield instruction is meant for this kind of use cases. So even if isb is slightly better on today's silicon, yield is likely to be better in the long run.

`SpinPause` should introduce short delays ~30 - 100 clocks. It should have ARM yield or Intel pause semantics to reduce the power consumption. As it is unlikely `yield` to become a real instruction in the long run, not `nop`. We tried different approaches to simulate `yield`. We found that `isb` is the closest to the needed semantics. 

Also without reviewing your benchmarking results, it is difficult to say what you measured. Could you share them?

You might also try another implementation: https://github.com/openjdk/jdk/pull/5562#issuecomment-1122652981

**Comment by eastig** [comment]
According to https://github.com/openjdk/jdk/pull/5562#issuecomment-966230710, `isb` works well on Apple M1.

**Comment by eastig** [comment]
@shipilev wrote on JDK-8321371:

> I remember trying the same thing along with [JDK-8318986](https://bugs.openjdk.org/browse/JDK-8318986), but I realized SpinPause() is only used from the quite hot VM native code, and so it would probably affect GC and runtime performance. The actual Thread.onSpinWait from Java code should be already handled by intrinsics. I suspect the overhead of doing the stub call is already similar to whatever hint we finally emit in the stub, but the WX transition back and forth is likely to be quite bad to make often. SpinWait is quite likely used in busy loops, so this would add up.

I don't know what WX transition is. 
I can comment:
> The actual Thread.onSpinWait from Java code should be already handled by intrinsics.

As bsd_aarch64 does not define what to use for spin wait, the intrinsic has no code.

**Comment by eastig** [comment]
The discussion of the ISB based apporach: RFC: [AArch64: Implementing spin pauses with ISB](https://mail.openjdk.org/pipermail/hotspot-dev/2021-August/054033.html)

**Comment by eastig** [comment]
> ... just hard code an isb instruction (or any other instruction that people can agree upon) like it's done on x86, that would probably be best.

I wanted the same but it was decided it'd be nice to have other options for spin wait/pause: `nop` and `yield`.

**Comment by fbredber** [comment]
I understand. Here the discussion went something like this:

The BSD port is almost strictly a MacOSX port. The configurability is more costly on MacOSX, because of the need to call os::current_thread_enable_wx().

The vast majority of MacOSX users don't want to configure which SpinPause instruction to use, they want something that is good straight out of the box.

Since there is a dedicated instruction (yield) in the AArch64 for this use case, we ought to use it. Because it's in the interest of the CPU vendor to have as good yield implementation as possible for each and every variety of the CPU. So if the user upgrades to a new CPU version, the SpinPause should still perform as good as possible, without the need to reconfigure.

For more info about the WX stuf, see here:
https://developer.apple.com/documentation/apple-silicon/porting-just-in-time-compilers-to-apple-silicon

**Comment by eastig** [comment]
> I understand. Here the discussion went something like this:
> 
> The BSD port is almost strictly a MacOSX port. The configurability is more costly on MacOSX, because of the need to call os::current_thread_enable_wx().
> 

Am I correct that in our current implementation if we call code of any generated stub we need to use `MACOS_AARCH64_ONLY(ThreadWXEnable wx(WXExec, thread));`?


> The vast majority of MacOSX users don't want to configure which SpinPause instruction to use, they want something that is good straight out of the box.

And they don't need to do. The configuration feature is for us to define the best spin pause implementation for an AArch64 CPU.
Different AArch64 CPU might need different implementations. For Neoverse CPUs and Apple M1, one ISB is good.

As I wrote if you don't choose an instruction for spin pause, you will have an empty onSpinWait intrinsic.

> 
> Since there is a dedicated instruction (yield) in the AArch64 for this use case, we ought to use it. 

`yield` first appeared in ARMv6. I think it can be considered as a legacy instruction. I don't know any currently existing ARM CPU which implements it. As it is not implemented, it is a NOP. 

> Because it's in the interest of the CPU vendor to have as good yield implementation as possible for each and every variety of the CPU.

It looks like vendors have not been interested in`yield` for 22 years since ARMv6 introduced in 2001.  Instead in Armv8.7-A and Armv9.2-A WFE/WFI with timeouts were introduced. They might allow more effective spin loops in hardware independent way. Cortex-X4, A720 and A520 support them. Implementations of Cortex-X4 are MediaTek Dimensity 9300 and Qualcomm Snapdragon 8 Gen 3. BTW, Apple M3 is ARMv8.6-A.

>  So if the user upgrades to a new CPU version, the SpinPause should still perform as good as possible, without the need to reconfigure.

We usually know ahead of a user, when a widely used CPU family gets new features useful for JVM. By the time users get a new CPU, they will get JVM supporting the new hardware features. With the current JDK Project release model, it's more likely users get them as soon as possible.

**Comment by eastig** [comment]
@fbredber 
I think what we have here is the XY problem. Please help me to understand what a problem you are trying to solve.

**Comment by fisk** [comment]
Let’s zoom out and look at the big picture for a bit.

So yield is the obvious instruction dedicated for this purpose in the ISA, and has been for a very long time. I suppose early ARM chips didn’t have a whole lot of concurrency and implementing it wasn’t all that beneficial as you frankly didn’t spend considerable time spinning. And now we seemingly have come to a classic chicken and egg problem. Software people like us don’t want to use the obvious yield instruction intended for this exact purpose, because hardware vendors haven’t implemented it. And hardware vendors don’t want to implement it, because no software is using it.

It feels like we had a sort of similar situation with neon vs SVE. All hardware was running neon, and nobody was running SVE. That doesn’t make it very encouraging as a software developer to implement SVE support in software, for an imaginary chip that doesn’t exist. And the fact that software doesn’t implement SVE doesn’t make it very encouraging to implement it. Yet we did it because it was the right thing to do, and no benchmark thanked us for it.

In the short term, it would seem like a better idea to use ISB, if you look at micro benchmarks. But what we are doing then is IMO what we tell our Java users not to do, and for good reason. We say “don’t use Unsafe to expose JDK and JVM internals that you happen to know how it works today, but you don’t know how it will work tomorrow, so you can look like a winner in a microbenchmark”. We are looking past the intended ISA contract, and look at current implementations today, and finding that as a hack, the ISB instruction which was designed to deal with cross modifying code, and not at all designed for this, is currently a better fit for doing what the yield instruction should have done had it been implemented, as the current ISB implementation has a long latency. And then a couple of years later, when the next LTS is released, a the Apple M3 is released. And then by the time we hit bulk of mainstream adoption, maybe we cycle through M4 and M5, and perhaps we end up with a hyper threaded core with 4 threads per core, and the isb turns out to be a disaster as it impedes progress of the other threads on the core to avoid shared resources being tripped over by cross modifying code, which is the exact opposite behaviour of what a pause instruction should do. Or maybe it won’t happen and everything is fine. The point is that *we don’t know*.

We are now in a situation where on MacOS the spinning hint does nothing. There is no cost over the baseline of making that spinning hint bind to the obvious yield instruction. It might not do a lot on Apple silicon today, and as of today, isb would probably yield better results (pun intended). But we don’t have to choose ISB just because it looks better today. We can instead be bold and break the chicken and egg situation, and use yield. Then the ball is on the HW court to do the right thing.

Because of this, I still think the right thing to do, is to bind SpinPause to the obvious ISA intended “yield” instruction, even though it is currently a nop.

**Comment by fbredber** [comment]
@eastig
> I think what we have here is the XY problem. Please help me to understand what a problem you are trying to solve.

I'd like to have a `SpinPause()` implementation in place for AArch64 based MacOSX platforms when I integrate https://github.com/openjdk/jdk/pull/17050, which removes `ObjectMonitor::NotRunnable()`.

**Comment by eastig** [comment]
@fisk I don't mind to use `yield`.

`isb` appeared as a result of real customer cases and consultations with hardware engineers. It was a real problem to write a micro-benchmark. I think it's still a problem.
We are aware that different implementations of AArch64 might need different spin pause implementations. This is why we provide options which instructions to use: `yield`, `nop` or `isb`.

I think hardcoding `yield` is not a good idea. We should use the current implemented approach where `yield` is selected. If the current approach does not work because it is based on a generated stub in CodeCache. Let's modify it to get code in place instead of stub.

@fbredber @fisk 
What do  you think about the following solution:

In `src/hotspot/cpu/aarch64/spin_wait_aarch64.hpp` we add `DEFAULT` to `SpinWait::Inst` which is set to `YIELD`. We also add a template function `exec<SpinWait::Instr>()` and its implementation `exec<SpinWait::YIELD>()`

```
class SpinWait {
public:
  enum Inst {
    NONE = -1,
    NOP,
    ISB,
    YIELD,
    DEFAULT = YIELD
  };
...
};

template<SpinWait::Inst inst> void exec();

template<> exec<YIELD>() {
  asm volatile("yield" : : : "memory");
}
```

In `src/hotspot/os_cpu/bsd_aarch64/os_bsd_aarch64.cpp`:

```
extern "C" {
  int SpinPause() {
    exec<SpinWait::DEFAULT>();
    return 1;
  }
```

You can also need to set `yield` as default in `src/hotspot/cpu/aarch64/globals_aarch64.hpp`
```
  product(ccstr, OnSpinWaitInst, "yield", DIAGNOSTIC,                    \
          "The instruction to use to implement "                        \
          "java.lang.Thread.onSpinWait()."                              \
          "Options: none, nop, isb, yield.")                            \
```

**Comment by fisk** [comment]
@eastig I think the essence of what you are proposing - having a non-WX solution for making the instruction selection configurable, but defaulting to yield - sounds very reasonable. Good suggestion!

**Comment by fbredber** [comment]
Sounds like a way forward. I'll try it out. Thanks!

**Comment by fbredber** [comment]
@eastig @fisk 

I've now rewritten `SpinPause()` so you can use the `OnSpinWaitInst` option to choose between `none`, `nop`, `isb` and `yield`.
I choose not to change the default from `none` to `yield` in this PR. For that I've created [JDK-8322535](https://bugs.openjdk.org/browse/JDK-8322535).

**Comment by eastig** [comment]
@fbredber 
The assembly code looks good to me.
Out of curiosity, why not to use a regular `switch`? Why is the manually written assembly better?
What if it looked like the following:
```
class SpinWait {
public:
   static void exec(Inst inst) {
      switch (inst) {
        case NOP:
            asm volatile();
            break;
        case ISB:
            asm volatile();
            break;
         case YIELD:
            asm volatile();
            break;
        default:
      }
   }
...
};

extern "C" {
  int SpinPause() {
    SpinWait::exec(VM_Version::spin_wait_desc().inst());
    return 1;
  }
```

**Comment by fbredber** [comment]
@eastig
> Out of curiosity, why not to use a regular `switch`? Why is the manually written assembly better?

I just like to keep away from conditional branches in code that is supposed to be in tight loops. :)

**Comment by dcubed-ojdk** [comment]
@fbredber - This PR's description and the final comment in [JDK-8321371](https://bugs.openjdk.org/browse/JDK-8321371)
are out of date relative to the change in direction. Please update the PR's description
and add a new comment to the bug to reflect the current design of the change.

File: src/hotspot/os_cpu/bsd_aarch64/os_bsd_aarch64.cpp [hunk 1]
```diff
@@ -514,7 +514,39 @@ static inline void atomic_copy64(const volatile void *src, volatile void *dst) {
...
+    uint64_t br_dst;
+    const int instructions_per_case = 2;
+    int64_t off = VM_Version::spin_wait_desc().inst() * instructions_per_case * Assembler::instruction_size;
+
+    assert(VM_Version::spin_wait_desc().inst() >= SpinWait::NONE &&
+           VM_Version::spin_wait_desc().inst() <= SpinWait::YIELD, "should be");
+    assert(-1 == SpinWait::NONE,  "should be");
+    assert( 0 == SpinWait::NOP,   "should be");
+    assert( 1 == SpinWait::ISB,   "should be");
+    assert( 2 == SpinWait::YIELD, "should be");
```
Comment (dcubed-ojdk) [inline src/hotspot/os_cpu/bsd_aarch64/os_bsd_aarch64.cpp:531]: nit: these should all be "must be" in keeping with the existing HotSpot style
for that type of assert(). In specification language, "should" is optional and
"must" is a requirement.

File: src/hotspot/os_cpu/bsd_aarch64/os_bsd_aarch64.cpp [hunk 2]
```diff
@@ -514,7 +514,39 @@ static inline void atomic_copy64(const volatile void *src, volatile void *dst) {
...
+
+    asm volatile(
+        "  adr  %[d], 20          \n" // 20 == PC here + 5 instructions => adress
```
Comment (dcubed-ojdk) [inline src/hotspot/os_cpu/bsd_aarch64/os_bsd_aarch64.cpp:534]: nit typo: s/adress/address/

**Review by dcubed-ojdk — APPROVED** [review]
Thumbs up with a couple of minor nits.

I like the idea of hooking the os_bsd_aarch64 version of `SpinPause`
into the OnSpinWaitInst option. This will enable easier experimentation
in the future without specially built binaries to switch the `SpinPause`
implementation.

File: src/hotspot/os_cpu/bsd_aarch64/os_bsd_aarch64.cpp (hunk 2 above)
Comment (fbredber) [reply src/hotspot/os_cpu/bsd_aarch64/os_bsd_aarch64.cpp:534]: My bad, but now you know how we spell it in Sweden. I'll fix.

**Comment by fbredber** [comment]
@dcubed-ojdk 
> Please update the PR's description and add a new comment to the bug to reflect the current design of the change.

Fixed

**Review by dcubed-ojdk — APPROVED** [review]
Thumbs up.

**Review by eastig — APPROVED** [review]
lgtm

**Comment by fbredber** [comment]
Thanks for the review guys. Since it's now the night before my Christmas vacation, I will wait and integrate this PR another year.

**Review by shipilev — APPROVED** [review]
All right, this one is pretty wild, but I somehow like it.

**Pull request closed without merging**